tree.R

```diff
@@ -165,112 +176,211 @@ build_tree <- function(df, filter_prevalence, filter_mean_abundance) {
+          function(node) node$name
+        ),
+        collapse = ", "
+      )
+    ))
     node$winner <- TRUE
     return()
   }
-  
+
   # drop from transposed data all correlated children
   transposed <- transposed %>%
     dplyr::select(., -dplyr::any_of(correlated_ids))
-  
+
   # run the random forest on the remaining parent + descendants
   # TODO: do the baked-in values need to be modifiable?
-  rf_winners <- rf_competition(transposed, metadata,"feature_of_interest", "subject_id", feature_type, sample_fraction, ncores, nperm)
-  
+  rf_winners <- rf_competition(transposed, metadata, "feature_of_interest", "subject_id", feature_type, sample_fraction, ncores, nperm)
+
+  # generate winner and loser name lists from the competitors (parent and non-correlated descendants), using the outcome
+  # build ahead of time so that a summary can be provided in outcomes
+  # this can be sped up and done in a single loop if outcomes are not needed
+  competitors <- append(not_correlated_descendant_winners, node)
+  winner_names <- list()
+  loser_names <- list()
+  for (competitor in competitors) {
+    if (competitor$id %in% rf_winners) {
+      winner_names <- append(winner_names, competitor$name)
+    } else {
+      loser_names <- append(loser_names, competitor$name)
+    }
+  }
+  outcome_str <- sprintf("winners: %s; losers: %s", paste(winner_names, collapse = ","), paste(loser_names, collapse = ","))
+
+  # now actually mark the results, including outcome string
   # mark winners/losers of parent and descendants
   # also mark the non-winners as rf losers
-  for (descendant in descendant_winners) {
-    if (descendant$id %in% rf_winners) {
-      descendant$winner <- TRUE
+  for (competitor in competitors) {
+    if (competitor$id %in% rf_winners) {
+      competitor$outcomes <- append(competitor$outcomes, sprintf("win: rf winner, %s", outcome_str))
+      competitor$winner <- TRUE
     } else {
-      descendant$winner <- FALSE
-      descendant$lost_rf <- TRUE
+      competitor$outcomes <- append(competitor$outcomes, sprintf("loss: rf loser, %s", outcome_str))
+      competitor$winner <- FALSE
+      competitor$lost_rf <- TRUE
     }
   }
-  
+
   return()
 }
 
+# compete all winners, updating the tree in the process
+# TODO: combine the overlaps in this code with the code in the function above
+compete_all_winners <- function(tree, metadata, sample_fraction, ncores) {
+  # all vs all competition with winners
+  # skipped rows have winner = FALSE so won't appear in this list
+  competitors <- get_descendant_winners(tree, tree$height)
+  if (length(all_winners) == 0) {
+    return(tree)
+  }
+
+  # all winners into a transposed data frame
+  df <- data.frame()
+  for (winner in competitors) {
+    df <- rbind(df, winner$abundance)
+  }
+  transposed <- as.data.frame(t(df))
+
+  # compete here
+  # return list of winner ids
+  # right now does nothing, all are marked as winners
+  winners <- lapply(competitors, function(node) node$id)
```

Comment:
this is where the all vs all competition function would go that returns winner ids, currently returns all ids so they all stay winners (nothing happens)